lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -1521,258 +1689,84 @@ LogicalResult FIRRTLLowering::visitDecl(MemOp op) {
-  Value reg =
-      builder.create<sv::RegOp>(rtl::UnpackedArrayType::get(dataType, depth),
-                                builder.getStringAttr(memName.str()));
-
-  // Some helpers.
-  auto buildPAssign = [&](Value dest, Value value) {
-    builder.create<sv::PAssignOp>(dest, value);
-  };
-
-  // Track pipeline registers that were added. These need to be
-  // randomly initialized later.
-  SmallVector<Value> pipeRegs;
+  FirMemory memSummary = analyzeMemOp(op);
 
   // Process each port in turn.
-  SmallVector<std::pair<Identifier, MemOp::PortKind>> ports = op.getPorts();
-
-  for (size_t i = 0, e = ports.size(); i != e; ++i) {
-    auto portName = ports[i].first.str();
-    auto port = op.getResult(i);
-
-    // Do not lower ports if they aren't used.
-    if (port.use_empty())
-      continue;
-
-    auto portBundleType =
-        port.getType().cast<FIRRTLType>().getPassiveType().cast<BundleType>();
-
-    // Create wires for all the memory ports
-    llvm::StringMap<Value> portWires;
-    for (BundleType::BundleElement elt : portBundleType.getElements()) {
-      auto fieldType = lowerType(elt.type);
-      if (fieldType.isInteger(0)) {
-        portWires.insert({elt.name.getValue(), Value()});
-        continue;
-      }
-      auto name =
-          (Twine(memName) + "_" + portName + "_" + elt.name.getValue()).str();
-      auto fieldWire = createTmpWireOp(fieldType, name);
-      portWires.insert({elt.name.getValue(), fieldWire});
-    }
-
-    // Now that we have the wires for each element, rewrite any subfields to
-    // use them instead of the subfields.
-    while (!port.use_empty()) {
-      auto portField = cast<SubfieldOp>(*port.user_begin());
-      portField->dropAllReferences();
-      (void)setLowering(portField, portWires[portField.fieldname()]);
-    }
-
-    // Don't emit anything more for zero bit data values.
-    if (dataType.isInteger(0))
-      continue;
+  SmallVector<Type, 8> resultTypes;
+  SmallVector<Value, 8> readOperands;
+  SmallVector<Value, 8> writeOperands;
+  DenseMap<Operation *, size_t> returnHolder;
+
+  size_t readCount = 0;
+  size_t writeCount = 0;
+  size_t readwriteCount = 0;
+
+  // Memories return multiple structs, one for each port, which means we have
+  // two layers of type to split appart.
+  for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+    auto portName = op.getPortName(i).getValue();
+    auto portKind = *op.getPortKind(i);
+
+    auto &portKindNum =
+        portKind == MemOp::PortKind::Read
+            ? readCount
+            : portKind == MemOp::PortKind::Write ? writeCount : readwriteCount;
+    auto addInput = [&](SmallVectorImpl<Value> &operands, StringRef field,
+                        size_t width) {
+      auto portType = IntegerType::get(op.getContext(), width);
+      auto accesses = getAllFieldAccesses(op.getResult(i), field);
+
+      Value wire = createTmpWireOp(
+          portType, ("." + portName + "." + field + ".wire").str());
```

Comment:
Should this use `.` or `_`? I thought we had some problems with `.` printing as `U23` or something?

Reply:
I was keeping the convention of temporary wire names, for which there were some strong opinions on keeping as ".".  As I don't have strong opinions on that...

Reply:
Not blocking. I'm fine with it as is. :+1: